source/loader/layers/sanitizer/asan_interceptor.cpp

```diff
@@ -0,0 +1,594 @@
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ * @file ur_sanitizer_layer.cpp
+ *
+ */
+
+#include "asan_interceptor.hpp"
+#include "device_sanitizer_report.hpp"
+#include "ur_sanitizer_layer.hpp"
+
+namespace ur_sanitizer_layer {
+
+namespace {
+
+// These magic values are written to shadow for better error
+// reporting.
+const int kUsmDeviceRedzoneMagic = (char)0x81;
+const int kUsmHostRedzoneMagic = (char)0x82;
+const int kUsmSharedRedzoneMagic = (char)0x83;
+const int kMemBufferRedzoneMagic = (char)0x84;
+const int kUnkownRedzoneMagic = (char)0x8F;
+
+const auto kSPIR_AsanShadowMemoryGlobalStart = "__AsanShadowMemoryGlobalStart";
+const auto kSPIR_AsanShadowMemoryGlobalEnd = "__AsanShadowMemoryGlobalEnd";
+
+const auto kSPIR_DeviceSanitizerReportMem = "__DeviceSanitizerReportMem";
+
+DeviceSanitizerReport SPIR_DeviceSanitizerReportMem;
+
+// uptr MemToShadow_CPU(uptr USM_SHADOW_BASE, uptr UPtr) {
+//     return USM_SHADOW_BASE + (UPtr >> 3);
+// }
+
+uptr MemToShadow_PVC(uptr USM_SHADOW_BASE, uptr UPtr) {
+    if (UPtr & 0xFF00000000000000ULL) { // Device USM
+        return USM_SHADOW_BASE + 0x200000000000ULL +
+               ((UPtr & 0xFFFFFFFFFFFFULL) >> 3);
+    } else { // Only consider 47bit VA
+        return USM_SHADOW_BASE + ((UPtr & 0x7FFFFFFFFFFFULL) >> 3);
+    }
+}
+
+// uptr MemToShadow_DG2(uptr USM_SHADOW_BASE, uptr UPtr) {
+//     if (UPtr & (~0xFFFFFFFFFFFFULL)) { // Device USM
+//         return USM_SHADOW_BASE + ((UPtr & 0xFFFFFFFFFFFFULL) >> 3);
+//     } else {
+//         return USM_SHADOW_BASE + (UPtr >> 3);
+//     }
+// }
+
+ur_context_handle_t getContext(ur_queue_handle_t Queue) {
+    ur_context_handle_t Context;
+    auto Result = context.urDdiTable.Queue.pfnGetInfo(
+        Queue, UR_QUEUE_INFO_CONTEXT, sizeof(ur_context_handle_t), &Context,
+        nullptr);
+    assert(Result == UR_RESULT_SUCCESS);
+    return Context;
+}
+
+ur_device_handle_t getDevice(ur_queue_handle_t Queue) {
+    ur_device_handle_t Device;
+    auto Result = context.urDdiTable.Queue.pfnGetInfo(
+        Queue, UR_QUEUE_INFO_DEVICE, sizeof(ur_device_handle_t), &Device,
+        nullptr);
+    assert(Result == UR_RESULT_SUCCESS);
+    return Device;
+}
+
+ur_program_handle_t getProgram(ur_kernel_handle_t Kernel) {
+    ur_program_handle_t Program;
+    auto Result = context.urDdiTable.Kernel.pfnGetInfo(
+        Kernel, UR_KERNEL_INFO_PROGRAM, sizeof(ur_program_handle_t), &Program,
+        nullptr);
+    assert(Result == UR_RESULT_SUCCESS);
+    return Program;
+}
+
+} // namespace
+
+ur_result_t SanitizerInterceptor::allocateMemory(
+    ur_context_handle_t Context, ur_device_handle_t Device,
+    const ur_usm_desc_t *Properties, ur_usm_pool_handle_t Pool, size_t Size,
+    void **ResultPtr, USMMemoryType Type) {
+    auto Alignment = Properties->align;
+    assert(Alignment == 0 || IsPowerOfTwo(Alignment));
+
+    auto ContextInfo = getContextInfo(Context);
+    auto DeviceInfo = ContextInfo->getDeviceInfo(Device);
```

Comment:
add a comment here saying that `Device` is `nullptr` if `Type == USMMemoryType::HOST`. I know there's a comment like that in the header file, but I think it might be useful here as well.

Reply:
Make sense. Done